### PR TITLE
daemon, option: Fix vlan bpf bypass ids loading

### LINKED ISSUE #20173: helm vlan.bpfBypass ignored
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Using `--set bpf.vlanBypass=0` in Helm is ignored. The flag shows up in viper, but not in the runtime config.

`# cat viper-agent-config.yaml | grep vlan-bpf`
vlan-bpf-bypass: 0

`# cat agent-runtime-config.json | grep VLAN`
 "VLANBPFBypass": [],

### Cilium Version

Client: 1.11.5 b0d3140 2022-05-10T02:47:42+02:00 go version go1.17.9 linux/amd64
Daemon: 1.11.5 b0d3140 2022-05-10T02:47:42+02:00 go version go1.17.9 linux/amd64

### Kernel Version

Linux 5.17.8-051708-generic #202205151940 SMP PREEMPT Mon May 16 13:36:10 EDT 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.0", GitCommit:"4ce5a8954017644c5420bae81d72b09b735c21f0", GitTreeState:"clean", BuildDate:"2022-05-03T13:46:05Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.0", GitCommit:"4ce5a8954017644c5420bae81d72b09b735c21f0", GitTreeState:"clean", BuildDate:"2022-05-03T13:38:19Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by pippolo84**
When installing cilium-agent via Helm, the configuration options are loaded reading files.
Each file represents an entry in the ConfigMap.
Doing so, all the values are loaded as strings and therefore must be converted to the specific option value type later.
This conversion is carried out by the spf13/viper package itself.

Internally, viper tries to load values from different sources:

- flag
- env
- config file
- key/value store
- default

When loading values from a config file, unlike the flag code path, the conversion from string to the requested value type is carried out by the spf13/cast package.
The cast package does not support the conversion from string to slice of ints and the viper.GetIntSlice(...) function does not report any error if the conversion goes wrong.

As a temporary workaround, the bpf.vlanBypass option value type has been changed to a slice of strings, so to leave the conversion to a slice of ints to be implemented manually.

To avoid this workaround, two things are needed:

1) Support for converting a string to a slice of ints in the cast package: https://github.com/spf13/cast/pull/146
2) Fix the `validateConfigmap` function to check option values using the cast package with the proper type conversion, instead of relying on pflag `Set` method.

Fixes: #20173

### CODE REVIEW

**Review by joestringer — APPROVED** [review]
docs change LGTM

File: pkg/option/config.go [hunk 1]
```diff
@@ -2940,7 +2940,15 @@ func (c *DaemonConfig) Populate() {
 	c.EnableRuntimeDeviceDetection = viper.GetBool(EnableRuntimeDeviceDetection)
 	c.EgressMultiHomeIPRuleCompat = viper.GetBool(EgressMultiHomeIPRuleCompat)
 
-	c.VLANBPFBypass = viper.GetIntSlice(VLANBPFBypass)
+	VLANBPFBypassIDs := viper.GetStringSlice(VLANBPFBypass)
```
Comment (sayboras) [inline pkg/option/config.go]: nit: VLANBPFBypassIDs -> vlanBPFBypassIDs

File: Documentation/configuration/vlan-802.1q.rst [hunk 2]
```diff
@@ -14,7 +14,7 @@ Cilium enables firewalling on native devices in use and will filter all unknown
 will always be passed through their main device with associated tag (e.g. VLAN device is ``eth0.4000`` and its main interface is ``eth0``).
 By default, Cilium will allow all tags from the native devices (i.e. if ``eth0.4000`` is controlled by Cilium and has
 an eBPF program attached, then VLAN tag ``4000`` will be allowed on device ``eth0``). Additional VLAN tags may be allowed
-with the cilium-agent flag ``--vlan-bpf-bypass=4001,4002`` (or Helm variable ``--set bpf.vlan-bpf-bypass=4001,4002``).
+with the cilium-agent flag ``--vlan-bpf-bypass=4001,4002`` (or Helm variable ``--set bpf.vlanBypass="4001 4002"``).
```
Comment (gandro) [inline Documentation/configuration/vlan-802.1q.rst]:
```suggestion
with the cilium-agent flag ``--vlan-bpf-bypass=4001,4002`` (or Helm variable ``--set bpf.vlanBypass="{4001,4002}"``).
```

**Review by gandro — COMMENTED** [review]
Awesome, one minor nit

File: Documentation/configuration/vlan-802.1q.rst (hunk 2 above)
Comment (sayboras) [inline Documentation/configuration/vlan-802.1q.rst]: I don't think the syntax for helm is no different with string value :thinking: , while this field is array type.

**Review by sayboras — COMMENTED** [review]
Couple of small comments as per below.

This is the only one we are using int slice, and unlikely we have int slice in future, so up to you if you want to add something similar to https://github.com/cilium/cilium/blob/master/contrib/scripts/check-viper-get-string-map-string.sh

File: Documentation/configuration/vlan-802.1q.rst (hunk 2 above)
Comment (sayboras) [reply Documentation/configuration/vlan-802.1q.rst]: oh ignore this comment, as Sebastian already highlighted the correct syntax.

**Review by kaworu — APPROVED** [review]
Thanks! Helm related changes LGTM :+1:

**Comment by pippolo84** [comment]
> Couple of small comments as per below.
> 
> This is the only one we are using int slice, and unlikely we have int slice in future, so up to you if you want to add something similar to https://github.com/cilium/cilium/blob/master/contrib/scripts/check-viper-get-string-map-string.sh

Thank you for the review!
Nice advice, I think it would be useful to add such a check. While adding a slice of ints option may be rare, it can lead to confusing and hard to debug problems.

File: daemon/cmd/daemon_main.go [hunk 3]
```diff
@@ -1094,7 +1094,7 @@ func initializeFlags() {
 	flags.Bool(option.ExternalClusterIPName, false, "Enable external access to ClusterIP services (default false)")
 	option.BindEnv(option.ExternalClusterIPName)
 
-	flags.IntSlice(option.VLANBPFBypass, []int{}, "List of explicitly allowed VLAN IDs, '0' id will allow all VLAN IDs")
```
Comment (tommyp1ckles) [inline daemon/cmd/daemon_main.go:1097]: Would it make sense to add a comment briefly describing this workaround and a link to the your PR?

**Review by tommyp1ckles — APPROVED** [review]
@pippolo84 Kubernetes changes LGTM

File: daemon/cmd/daemon_main.go (hunk 3 above)
Comment (pippolo84) [reply daemon/cmd/daemon_main.go:1097]: Sure, it may be useful. I'll fix it, thank you!

File: contrib/scripts/check-viper-unusable-api.sh [hunk 4]
```diff
@@ -13,3 +15,10 @@ if grep -r --exclude-dir={.git,_build,vendor,contrib} -i --include \*.go "String
   echo "Found flags.StringToStringVar usage. Please use option.NewNamedMapOptions instead";
   exit 1
 fi
+
+# viper.GetIntSlice
+# Related Cilium issue https://github.com/cilium/cilium/issues/20173 and companion PR https://github.com/cilium/cilium/pull/20282
+if grep -r --exclude-dir={.git,_build,vendor,contrib} -i --include \*.go "viper.GetIntSlice" .; then
+  echo "Found viper.GetIntSlice(key) usage. Please use a flags.StringSlice type and viper.GetStringSlice(key) instead";
+  exit 1
+fi
```
Comment (sayboras) [inline contrib/scripts/check-viper-unusable-api.sh]: nit: new line

**Review by sayboras — APPROVED** [review]
LGTM :100:

**Review by gandro — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by sayboras** [comment]
/test-1.24-net-next

**Review by christarazi — APPROVED** [review]

**Review by rolinh — COMMENTED** [review]
LGTM but given that this PR will be backported to v1.11, shouldn't the breaking change (`--set bpf.vlan-bpf-bypass=4001,4002` => `--set bpf.vlanBypass="{4001,4002}"`) be documented in the upgrade notes?

**Comment by gandro** [comment]
> LGTM but given that this PR will be backported to v1.11, shouldn't the breaking change (`--set bpf.vlan-bpf-bypass=4001,4002` => `--set bpf.vlanBypass="{4001,4002}"`) be documented in the upgrade notes?

It's not a breaking change. The old syntax just  simply never worked, it was a complete no-op.
The actual breaking change happened in https://github.com/cilium/cilium/pull/20304 which is not backported to v1.11.

**Review by rolinh — APPROVED** [review]

**Review by ldelossa — APPROVED** [review]

**Pull request merged**